core/src/main/scala/org/apache/spark/internal/config/package.scala

```diff
@@ -1391,10 +1391,12 @@ package object config {
 
   private[spark] val SHUFFLE_HOST_LOCAL_DISK_READING_ENABLED =
     ConfigBuilder("spark.shuffle.readHostLocalDisk")
-      .doc(s"If enabled (and `${SHUFFLE_USE_OLD_FETCH_PROTOCOL.key}` is disabled and external " +
-        s"shuffle `${SHUFFLE_SERVICE_ENABLED.key}` is enabled), shuffle " +
-        "blocks requested from those block managers which are running on the same host are read " +
-        "from the disk directly instead of being fetched as remote blocks over the network.")
+      .doc(s"If enabled (and `${SHUFFLE_USE_OLD_FETCH_PROTOCOL.key}` is disabled and 1) external " +
+        s"shuffle `${SHUFFLE_SERVICE_ENABLED.key}` is enabled or 2) ${DYN_ALLOCATION_ENABLED.key}" +
+        s" is disabled), shuffle blocks requested from those block managers which are running on " +
```

Comment:
Why does dynamic allocation need to be disabled?

Reply:
Good point. It is true that in case of dynamic allocation executors come and go but an executor can be lost without dynamic allocation too and the code is prepared to handle those: it leads to a FailureFetchResult so I expect in case of errors we are as good as before. On the other hand this still helps when local block successfully read.

So I cannot see the why dynamic allocation is excluded from this feature.

Reply:
Yes, it prevents the case where executors could come and go in dynamic allocation. Also, I think it's still different from executor loss error. Because executor loss is an abnormal case which out of control of Spark while dynamic allocation is under control. And executor shutdown in dynamic allocation happens more frequently compares to executor loss. I think we should try our best to avoid shuffle fetch failure since its penalty is not trivial, especially when we can avoid it.

Besides, for the case of dynamic allocation enabled, users could already use external shuffle service. Therefore, I can't think of a strong reason to mix these two branches.

P.S. we could probably allow dynamic allocation here if `spark.dynamicAllocation.shuffleTracking.enabled` is also enabled.

Reply:
As of Spark 3 we no longer require dynamic allocation to have a shuffle service. Can you explain what would fail in the current approach with dynamic allocation which requires disabiling it? And also document that it does not function with dynamic allocation.

Reply:
> As of Spark 3 we no longer require dynamic allocation to have a shuffle service.

This exactly what I mentioned in P.S.. I can update the documentation according to this feature.

But please also note dynamic allocation without external shuffle service is still an experimental feature disabled by default. And it has a main problem that the user needs to config when to delete shuffle files while most common users have no idea about this. And by default, shuffle files will not be removed until GC happens at the driver side. It also means executors won't come and go more frequently than dynamic allocation with shuffle service. Therefore, I think we were discussing a more general problem above when using dynamic allocation with shuffle service.

Reply:
I still don’t see what would not work with this feature, maybe you can try and explain it in another manner?

Reply:
I mean, it *can* work with this feature enabled(please see my P.S. comment above). And I will update the document about dynamic allocation to:

1. not allow when `spark.dynamicAllocation.enabled=true` && `spark.dynamicAllocation.shuffleTracking.enabled=false`

2. allow when `spark.dynamicAllocation.enabled=true` && `spark.dynamicAllocation.shuffleTracking.enabled=true`

also cc @attilapiros Does it make sense to you?